Start shipping the Cargo book

steveklabnik opened the pull request:
Fixes #44910
Fixes #39588

See both of those bugs for more details.

rust-highfive commented:
r? @alexcrichton

(rust_highfive has picked a reviewer for you, use r? to override)

alexcrichton commented on src/bootstrap/doc.rs:
```diff
@@ -133,6 +133,49 @@ impl Step for UnstableBook {
+        run.builder.ensure(CargoBook {
+            target: run.target,
+            name: INTERNER.intern_str("cargo"),
+            src: INTERNER.intern_path(PathBuf::from("src/tools/cargo/src/doc/book")),
```
I think this and the `name` field above are best left as local variables in `run` perhaps because I don't think they're inputs so much as calculated.

steveklabnik replied:
ah makes sense, I did a bit TOO much cargo culting.

I'll update this after we decide what the name should be, and then :shipit:

behnam commented on src/doc/index.md:
```diff
@@ -28,6 +28,7 @@ Rust provides a number of book-length sets of documentation, collectively
 nicknamed 'The Rust Bookshelf.'
 
 * [The Rust Programming Language][book] teaches you how to program in Rust.
+* [The Cargo Book][cargo-book] is a guide to Cargo, Rust's build tool and dependency manager.
```
We're calling it "The Cargo Manual" in the cargo repo. Should we update that to say "The Cargo Book"?

steveklabnik replied:
I'm okay with either, but the rest are almost all called "Book". I'll defer to @rust-lang/cargo here

alexcrichton replied:
I'm fine with "The Cargo Book", it jives better with me with "bookshelf"

behnam replied:
Cool! Let's go with what you have here and I'll submit a PR for cargo to update the title there.

behnam replied:
PR to update cargo: https://github.com/rust-lang/cargo/pull/4695

steveklabnik commented:
@bors: r=alexcrichton

bors commented:
:pushpin: Commit ac90301 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton ac90301c48fb4e1d1b05ac5b4be62be768b52ae5 -->

bors commented:
:hourglass: Testing commit ac90301c48fb4e1d1b05ac5b4be62be768b52ae5 with merge af64ce8e3be13f3bc1bed4021c25f1a4df74fc7d...

bors commented:
:broken_heart: Test failed - [status-travis](https://travis-ci.org/rust-lang/rust/builds/296629643?utm_source=github_status&utm_medium=notification)

kennytm commented:
`x86_64-gnu-distcheck` failed, broken links.

```
[01:39:55] cargo/print.html:97: broken link - cargo/getting-started.html
[01:39:55] cargo/print.html:790: broken link - cargo/config.html
[01:39:55] cargo/print.html:2671: broken link fragment `#the-dependencies-section` pointing to `cargo/reference/manifest.html`
[01:39:55] cargo/faq.html:144: broken link fragment `#the-dependencies-section` pointing to `cargo/reference/manifest.html`
[01:39:55] cargo/reference/specifying-dependencies.html:372: broken link - cargo/config.html
[01:39:55] cargo/index.html:96: broken link - cargo/getting-started.html
```

shepmaster commented:
Triage ping @steveklabnik — we haven't heard from you in a week and this PR has some test failures that need to be addressed!

steveklabnik commented:
Whoops, thanks. Will look into this soon.

behnam commented:
I checked these files (`cargo/print.html` and `cargo/faq.html`, at least) on the cargo repo, under `target/doc/book/` and all the links look fine when built. Not sure what's different in the rust build that could cause this. (Haven't tried building this PR on machine yet, because of some build issues.)

behnam commented:
Oh, I just saw that you fixed some of the links, @steveklabnik! Thanks!

steveklabnik commented:
Yeah, it's mostly due to some small issues.

steveklabnik commented:
@alexcrichton are we free to just update the cargo repo whenever? We'd need to do that to land this PR, I thought it was generally updated once a cycle. I could be wrong.

Mark-Simulacrum commented:
We're free to update the Cargo submodule whenever, yes. We try to do it at least once a cycle, but generally it happens about 3-5 times based on bugfixes landing.

steveklabnik commented:
Great! Tested this locally and it should pass now, though I had to rebase as cargo was also updated upstream; let's see what travis says!

alexcrichton commented:
@bors: r+

bors commented:
:pushpin: Commit 8f0f3b1 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton 8f0f3b131105b0723d4428b36118f4f9d32f91bf -->

bors commented:
:hourglass: Testing commit 8f0f3b131105b0723d4428b36118f4f9d32f91bf with merge 4ef9cb0494c4267c6f202d2b1a3832c00bdff5b5...

bors commented:
:broken_heart: Test failed - [status-travis](https://travis-ci.org/rust-lang/rust/builds/300752226?utm_source=github_status&utm_medium=notification)

kennytm commented:
FreeBSD, could not compile cargo because libc is updated. 

<details>

```
[01:20:28] [m[m[32m[1m   Compiling[m cargo v0.24.0 (file:///checkout/src/tools/cargo)
[01:20:30] [1m[31merror[E0308](B[m[1m: mismatched types(B[m
[01:20:30]    (B[m[1m[34m--> (B[m/checkout/src/tools/cargo/src/cargo/core/shell.rs:299:49(B[m
[01:20:30]     (B[m[1m[34m|(B[m
[01:20:30] [1m[34m299(B[m (B[m[1m[34m| (B[m            if libc::ioctl(libc::STDERR_FILENO, libc::TIOCGWINSZ, &mut winsize) < 0 {(B[m
[01:20:30]     (B[m[1m[34m| (B[m                                                (B[m[1m[31m^^^^^^^^^^^^^^^^(B[m (B[m[1m[31mexpected u64, found u32(B[m
[01:20:30] 
[01:20:34] [1m[31merror(B[m[1m: aborting due to previous error(B[m
[01:20:34] 
[01:20:34] [m[m[31m[1merror:[m Could not compile `cargo`.
[01:20:34] 
[01:20:34] To learn more, run the command again with --verbose.
[01:20:34] 
```

</details>

steveklabnik commented:
I... have no idea what to do here, then.

kennytm commented:
@steveklabnik Cargo needs to modify `src/cargo/core/shell.rs` to make it portable across unixes. Maybe just adding an `as` cast (or `.into()` if possible) is enough. After that, we update the sub module here again.

alexcrichton commented:
I've opened https://github.com/rust-lang/cargo/pull/4713 to hopefully fix this.

kennytm commented:
rust-lang/cargo#4713 has been merged, the cargo submodule can be updated again. @steveklabnik

bors commented:
:umbrella: The latest upstream changes (presumably #45949) made this pull request unmergeable. Please resolve the merge conflicts.

steveklabnik commented:
Should be resolved!

Mark-Simulacrum commented:
I feel like this shouldn't need to update Cargo -- perhaps that's accidental / needs a rebase? I think current Cargo master is in `src/tools/cargo` (or at least recent enough for this repo).

steveklabnik commented:
I had updated it because I needed to update before; I guess given the conflict, it's probably not needed. I've been travelling and so I haven't been paying as close attention to the queue as usual.

I've removed that bit and rebased, let's see what Travis says.

steveklabnik commented:
@bors: r=alexcrichton

looks like this is passing!

bors commented:
:pushpin: Commit 3b32a3a has been approved by `alexcrichton`

<!-- @bors r=alexcrichton 3b32a3a104732004ca1b9e40a38221dd1cf3eb7f -->

bors commented:
:hourglass: Testing commit 3b32a3a104732004ca1b9e40a38221dd1cf3eb7f with merge 42ea30c8b0107cced6c85f9240111d61aa3ff654...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.5373), [status-travis](https://travis-ci.org/rust-lang/rust/builds/302796042?utm_source=github_status&utm_medium=notification)
Approved by: alexcrichton
Pushing 42ea30c8b0107cced6c85f9240111d61aa3ff654 to master...

behnam commented:
For some reason the book is not correctly built on auto-deploy. I filed a new issue for it: <https://github.com/rust-lang/rust/issues/46195>
